File: 46-permutations/permutations.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();
        findPermutations(nums, 0, ans);
        return ans;
    }

    public void findPermutations(int[] nums, int ind, List<List<Integer>> ans) {
        if (ind == nums.length) {
            List<Integer> each = new ArrayList<>();
            for (int num : nums) {
                each.add(num);
            }
            ans.add(each);
            return;
        }

        for (int i = ind; i < nums.length; i++) {
            swap(nums, i, ind);
            findPermutations(nums, ind + 1, ans);
            swap(nums, i, ind); // backtrack
        }
    }

    public void swap(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
}

File: 46-permutations/permutations.cpp
class Solution {
public:
    vector<vector<int>> ans;
    vector<vector<int>> permute(vector<int>& nums) {
        
        // vector<int> sub;
          findSol(nums,0);

          return ans;
    }
    void findSol(vector<int>&nums,int ind){
          if(ind==nums.size()){
              ans.push_back(nums);
              return;
          }

          for(int i=ind;i<nums.size();i++){
             swap(nums[ind],nums[i]);
             findSol(nums,ind+1);
             swap(nums[ind],nums[i]);
          }
    }
};